relay-wstring/src/lib.rs

```diff
@@ -434,21 +439,58 @@ unsafe fn decode_surrogates(u: u16, u2: u16) -> char {
 }
 
 /// Checks that the raw bytes are valid UTF-16LE.
+///
+/// When an error occurs this code needs to set `error_len` to skip forward 2 bytes *if*
+/// there still are more bytes to consume.  This is the repetitive
+/// `chunks.next().map(|_chunk| std:mem::size_of::<u16>() as u8)`, but we compute this
+/// lazily for performance.
+///
+/// Likewise we compute `valid_up_to` lazily for performance.
 fn validate_raw_utf16le(raw: &[u8]) -> Result<(), Utf16Error> {
-    // This could be optimised as it does not need to be actually decoded, just needs to
-    // be a valid byte sequence.
-    if raw.len() % 2 != 0 {
-        return Err(Utf16Error::new());
-    }
-    let u16iter = raw
-        .chunks_exact(2)
-        // Using filter_map to avoid generating panicking code
-        .filter_map(|chunk| u16::copy_from_le_bytes(chunk));
-    if std::char::decode_utf16(u16iter).all(|result| result.is_ok()) {
-        Ok(())
-    } else {
-        Err(Utf16Error::new())
+    let base_ptr = raw.as_ptr() as usize;
+    let mut chunks = raw.chunks_exact(std::mem::size_of::<u16>());
+
+    while let Some(chunk) = chunks.next() {
+        let code_point_ptr = chunk.as_ptr() as usize;
+        // Chunks always returns 2 bytes, avoid generating panicking code
+        let u = u16::copy_from_le_bytes(chunk).unwrap_or_default();
+
+        if is_trailing_surrogate(u) {
+            return Err(Utf16Error {
+                valid_up_to: code_point_ptr - base_ptr,
+                error_len: chunks.next().map(|_chunk| std::mem::size_of::<u16>() as u8),
+            });
+        }
+
+        if is_leading_surrogate(u) {
+            match chunks.next().and_then(u16::copy_from_le_bytes) {
+                Some(u2) => {
+                    if !is_trailing_surrogate(u2) {
+                        return Err(Utf16Error {
+                            valid_up_to: code_point_ptr - base_ptr,
+                            error_len: chunks.next().map(|_chunk| std::mem::size_of::<u16>() as u8),
```

Comment:
The error len can statically be `2` here, since `u2` is a valid UTF-16 starting character. `None` is invalid here.

Imagine this sequence: `[single] - [single] - [leading] - [single]`. Our error should have `valid_up_to = 4` and `error_len = 2`, because you can parse a valid UTF-16 string for `6..8`.

Reply:
On rethinking this, you could also be more elaborate (although I'm not sure if that makes sense) and then look even further into the future, but you will have to differentiate based on what the next character is. 

However, the intent behind `error_len` is to print a box character. And in that sense, you may want to print another box character after this, if you, for instance, get another leading character immediately afterwards.

Reply:
hmm, there is no guarantee that `error_len` is only allowed to cover only one char is there?  Interestingly there *is* a maximum number of bytes in case of `None` according to the description.  So maybe there is this assumption and we can't optimise looking forward further.  The fact that `Utf8Error` stores their `error_len` in a `u8` also re-enforces this assumption i guess.

Reply:
> there is no guarantee that error_len is only allowed to cover only one char is there?

Looks like there is no such guarantee. However, when reading the description, it sounds like this should cover one replacement character. I'm assuming, for UTF-8 `error_len` will always be in `1..=3`.

Following up with my initial comment: The only edge case why `chunks.next().map()` is asking too much is because it would swallow the last code point in the above example. Imagine the above string renders as `AB�D` and the replacement character is due to a leading surrogate without its trailing counterpart. Your code would probably render as `AB` or `AB�`, since your Error has `valid_up_to=4` and `error_len=None` (due to chunks.next() being None).

Reply:
Agreed the intention is to be able to insert a replacement char for each error, so we can't skip ahead and combine errors.

So your original comment is right, a lone trailing surrogate can be hardcoded to `error_len=2`.  Also so for a lone leading surrogate.  So we can never return an `error_len` different that 2.  Only a lone leading surrogate at the end of the slice can return `None`.

I don't follow your last example, I think this would result in `AB�D`.